packages/happn-cluster/lib/services/proxy.js

```diff
@@ -90,24 +91,33 @@ Proxy.prototype.start = require('util').promisify(function (callback) {
 
   this.__proxyServer = proxy.createProxyServer(opts);
 
+  let done, returnError;
   var onError = function (error) {
     self.__proxyServer._server.removeListener('listening', onListening);
-    callback(error);
+    returnError(error);
   };
-
   var onListening = function () {
     self.log.info('forwarding %s to %s', listening, target);
     self.__proxyServer._server.removeListener('error', onError);
 
     self.__proxyServer.on('error', self.__onProxyErrorListener);
     self.__proxyServer._server.on('error', self.__onServerErrorListener);
-    callback();
+    if (self.config.port === 0) {
+      let proxyPort = self.__proxyServer._server.address().port;
+      self.config.port = proxyPort;
+      self.happn.config.services.proxy.port = proxyPort;
+    }
+    // console.log('ADDRESS: ', self.__proxyServer._server.address());
+    done();
   };
-
   this.__proxyServer.listen(port, host); // need to listen() first, it creates the internal _server (odd)
   this.__proxyServer._server.once('error', onError);
   this.__proxyServer._server.once('listening', onListening);
-});
+  return new Promise((res, rej) => {
```

Comment:
can we make more sense of this

Reply:
Not right now!! ??

Reply:
its fine I refactored the proxy to make more sense.